Deprecate BasePrimitiveResult

chriseclectic opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Deprecates the unnecessary base class BasePrimitiveResult so we can have a clean break in defining new ones in Qiskit 1.0. Also deprecates the common utility methods `decompose`, `experiments`, and `num_experimetns` this base class adds to SamplerResult and EstimatorResult.

### Details and comments

t-imamichi reviewed (dismissed):
LGTM

jakelishman commented on qiskit/primitives/base/base_result.py:
```diff
@@ -71,17 +76,38 @@ def _generate_experiments(self) -> Iterator[ExperimentData]:
+            category=DeprecationWarning,
+            stacklevel=2,
+        )
+        super().__post_init__()
```
Neither this class nor anything in its inheritance chain is a `dataclasses.dataclass`, so the `__post_init__` method will not get called if it is instantiated.  You might be better off deprecating retrieving the `BasePrimitiveResult` _name_ from all modules that it's publicly exported from.

chriseclectic replied:
I implemented @jakelishman suggestion to deprecate on attempting to import BasePrimitiveResult

jakelishman reviewed (commented):
Also, given Ian's RFC, is it actually set in stone that these _will_ be removed in the way they're being done here, or is there a chance that there will just be a completely disjoint versioned set of primitives APIs, in which case these needn't be so eagerly deprecated?

chriseclectic commented:
@t-imamichi Updated this PR if you could re-review

chriseclectic committed:
Deprecate BasePrimitiveResult

Deprecates the unnecessary base class BasePrimitiveResult so we can have a clean break in defining new ones in Qiskit 1.0. Also deprecates the common utility methods `decompose`, `experiments`, and `num_experimetns` this base class adds to SamplerResult and EstimatorResult.

Add jakes suggestion

t-imamichi reviewed (approved):
LGTM
